teps/0028-task-results-without-results.md

```diff
@@ -0,0 +1,358 @@
+any defaults and have to be initialized at the `pipeline` or `pipelineRun` level.
+
+The proposal is designed on the similar lines, when a `param` is declared `optional` at the `task` level and referring to the
+`task result` of any other task in a `pipeline`, is allowed to have a default at the `pipelineTask` level.
+
+In the following example, `pipelineTask` add declares `results` and its default value. Now, `x` is an `optional` param
+in `multiply` task and referring to a task result of `add`. If `add` fails for some reason, finally task `multiply`
+continues execution with default `sum` which is `0`:
+
+```yaml
+spec:
+  tasks:
+    - name: add
+      taskRef:
+        name: add-task
+      params:
+        - name: first
+          value: $(params.first)
+        - name: second
+          value: $(params.second)
+      results:
+        - name: sum
+          value: 0
+  finally:
+    - name: multiply
+      taskRef:
+        name: multiply
+      params:
+        - name: x
+          value: $(task.add.results.sum)
+```
+
+No default for a `required` task param (referring to a task result) allowed. If a task result is missing, `pipeline`
+skips that `pipelineTask`.
+
+### User Stories
+
+#### Story 1
+
+For a cleanup use case, `finally` task receives the project name, or the configuration name as a `task result` from
+the `acquire` task. If `acquire` task fails to acquire any resources, `cleanup` is skipped.
+
+#### Story 2
+
+Need user story for `finally` task with optional `param` referring to a task result.
+
+#### Story 3
+
+Need user story for a task with `continueAfterSkip` set to `true` with missing task result.
+
+#### Story 4
+
+Need user story for a when expression with missing task result.
+
+## Design Details
+
+### finally with missing task result
+
+`Pipeline` author can specify `default` for a missing result if that `param` is optional.
+
+```yaml
+spec:
+  params:
+    - name: owner-name
+      default: "the-best-owner"
+  tasks:
+   - name: boskos-acquire
+     taskRef:
+       name: boskos-acquire
+     params:
+       - name: owner-name
+         value: $(params.owner-name)
+     results:
+      - name: leased-resource
+        value: "fake-awesome-project"
+   - name: do-stuff-with-resource
+     taskRef:
+       name: do-stuff-with-resource
+     params:
+       - name: resource-name
+         value:  $(tasks.boskos-acquire.results.leased-resource)
+ finally:
+   - name: boskos-release
+     taskRef:
+       name: boskos-release
+     params:
+       - name: leased-resource
+         value:  $(tasks.boskos-acquire.results.leased-resource)
+```
+
+
+### when expression with missing task result
+
+```yaml
+spec:
+  tasks:
+    - name: add
+      taskRef:
+        name: add-task
+      params:
+        - name: first
+          value: $(params.first)
+        - name: second
+          value: $(params.second)
+      results:
+        - name: sum
+          value: 0
```

Comment:
> 1.

yup sure, lets call it `default` instead of `value`. Yes use this default value when a task does not produce a result. 

> 2. 

We have an issue to declare such task (does not produce a result) failure: https://github.com/tektoncd/pipeline/issues/3497. This fix will prevent executing any other dag tasks and no need for any defaults. But finally tasks are still executed and you suggest not using default in finally as well?